src/devices/src/virtio/net/device.rs

```diff
@@ -0,0 +1,147 @@
+use vm_virtio::Queue;
+
+use crate::virtio::features::{VIRTIO_F_IN_ORDER, VIRTIO_F_RING_EVENT_IDX, VIRTIO_F_VERSION_1};
+use crate::virtio::net::{Error, NetArgs, Result, NET_DEVICE_ID};
+use crate::virtio::{CommonConfig, Env, SingleFdSignalQueue, QUEUE_MAX_SIZE};
+
+use super::bindings;
+use super::features::*;
+use super::queue_handler::QueueHandler;
+use super::simple_handler::SimpleHandler;
+use super::tap::Tap;
+
+pub struct Net<M: GuestAddressSpace> {
+    cfg: CommonConfig<M>,
+    tap_name: String,
+}
+
+impl<M> Net<M>
+where
+    M: GuestAddressSpace + Clone + Send + 'static,
+{
+    pub fn new<B>(env: &mut Env<M, B>, args: &NetArgs) -> Result<Arc<Mutex<Self>>>
+    where
+        // We're using this (more convoluted) bound so we can pass both references and smart
+        // pointers such as mutex guards here.
+        B: DerefMut,
+        B::Target: MmioManager<D = Arc<dyn DeviceMmio + Send + Sync>>,
+    {
+        let device_features = (1 << VIRTIO_F_VERSION_1)
+            | (1 << VIRTIO_F_RING_EVENT_IDX)
+            | (1 << VIRTIO_F_IN_ORDER)
+            | (1 << VIRTIO_NET_F_CSUM)
+            | (1 << VIRTIO_NET_F_GUEST_CSUM)
+            | (1 << VIRTIO_NET_F_GUEST_TSO4)
+            | (1 << VIRTIO_NET_F_GUEST_UFO)
+            | (1 << VIRTIO_NET_F_HOST_TSO4)
+            | (1 << VIRTIO_NET_F_HOST_UFO);
+
+        // An rx/tx queue pair.
+        let queues = vec![Queue::new(env.mem.clone(), QUEUE_MAX_SIZE); 2];
+        // TODO: We'll need a minimal config space to support setting an explicit MAC addr
+        // on the guest interface at least. We use an empty one for now.
+        let config_space = Vec::new();
+        let virtio_cfg = VirtioConfig::new(device_features, queues, config_space);
+
+        let common_cfg = CommonConfig::new(virtio_cfg, env).map_err(Error::Generic)?;
+
+        let net = Arc::new(Mutex::new(Net {
+            cfg: common_cfg,
+            tap_name: args.tap_name.clone(),
+        }));
+
+        env.register_mmio_device(net.clone())
+            .map_err(Error::Generic)?;
+
+        Ok(net)
+    }
+}
+
+impl<M: GuestAddressSpace + Clone + Send + 'static> VirtioDeviceType for Net<M> {
+    fn device_type(&self) -> u32 {
+        NET_DEVICE_ID
+    }
+}
+
+impl<M: GuestAddressSpace + Clone + Send + 'static> Borrow<VirtioConfig<M>> for Net<M> {
+    fn borrow(&self) -> &VirtioConfig<M> {
+        &self.cfg.virtio
+    }
+}
+
+impl<M: GuestAddressSpace + Clone + Send + 'static> BorrowMut<VirtioConfig<M>> for Net<M> {
+    fn borrow_mut(&mut self) -> &mut VirtioConfig<M> {
+        &mut self.cfg.virtio
+    }
+}
+
+impl<M: GuestAddressSpace + Clone + Send + 'static> VirtioDeviceActions for Net<M> {
+    type E = Error;
+
+    fn activate(&mut self) -> Result<()> {
+        let rxq = self.cfg.virtio.queues[0].clone();
+        let txq = self.cfg.virtio.queues[1].clone();
+
+        let tap = Tap::open_named(self.tap_name.as_str()).map_err(Error::Tap)?;
```

Comment:
Doesn't this also creates the device if it isn't already present? The PR description mentions how to create tap devices in order for the tests to run, but shouldn't they be created anyway assuming that the user has `CAP_NET_ADMIN`?